In [1]:
# Importing Libraries
import selenium
import pandas as pd
import time
from bs4 import BeautifulSoup

# Importing selenium webdriver 
from selenium import webdriver

# Importing required Exceptions which needs to handled
from selenium.common.exceptions import StaleElementReferenceException, NoSuchElementException

#Importing requests
import requests

# importing regex
import re

In [2]:
# Activating the chrome browser
driver=webdriver.Chrome("chromedriver.exe") 
time.sleep(3)

# Opening the homepage of Amazon.in
url = "https://timesofindia.indiatimes.com/archive.cms"
driver.get(url)

time.sleep(2)

In [3]:
#finding feb 2018 button
feb2018_button = driver.find_element_by_xpath('//a[@class="normtxt"][36]')

In [4]:
#clicking on the button
feb2018_button.click()

In [5]:
#creating an empty list to store each days url link
url=[]

In [6]:
#since we have different tag name for each row of days . we extend aselecting multiple elements to get all url
url=driver.find_elements_by_xpath('//table[@id="calender"]/tbody/tr[2]/td/a')

In [7]:
url.extend(driver.find_elements_by_xpath('//table[@id="calender"]/tbody/tr[3]/td/a'))

In [8]:
url.extend(driver.find_elements_by_xpath('//table[@id="calender"]/tbody/tr[4]/td/a'))

In [9]:
url.extend(driver.find_elements_by_xpath('//table[@id="calender"]/tbody/tr[5]/td/a'))

In [10]:
url.extend(driver.find_elements_by_xpath('//table[@id="calender"]/tbody/tr[6]/td/a'))

In [11]:
#we find length of url is 32 in which first 4 url is of empty space. we can remove it.
len(url)

32

In [12]:
#concluding final urls
url=url[4:]

In [13]:
len(url)

28

In [14]:
#extracting each news link present in every url
news_link=[]

for i in url:
    news_link.append(i.get_attribute("href"))
    

In [15]:
#we got newslink for all days. I tried doing all at once but data was lost due to power failure and internet issue. So i divided news_link into 6 parts from feb1-5 to feb 26-28 . This helped in extracting data
news_link=news_link[0:5]

In [16]:
news link for feb1 - feb 5
news_link

['https://timesofindia.indiatimes.com/2018/2/1/archivelist/year-2018,month-2,starttime-43132.cms',
 'https://timesofindia.indiatimes.com/2018/2/2/archivelist/year-2018,month-2,starttime-43133.cms',
 'https://timesofindia.indiatimes.com/2018/2/3/archivelist/year-2018,month-2,starttime-43134.cms',
 'https://timesofindia.indiatimes.com/2018/2/4/archivelist/year-2018,month-2,starttime-43135.cms',
 'https://timesofindia.indiatimes.com/2018/2/5/archivelist/year-2018,month-2,starttime-43136.cms']

In [17]:
# All the news articles in the news link is extracted by
final=[]
for page in news_link:
    driver.get(page)
    time.sleep(10)
    
    news=driver.find_elements_by_xpath('//span[@style="font-family:arial ;font-size:12;color: #006699"]/a')
    
    for i in news:
        final.append(i.get_attribute("href"))

In [18]:
#There are 2500 news article found in each of the 5 days
len(final)

2500

In [19]:
#link for each article present
final

['http://timesofindia.indiatimes.com//city/kochi/celebrity-reporter-actress-ahaana-krishna-writes-about-the-need-for-including-moral-science-in-schools/articleshow/62724803.cms',
 'http://timesofindia.indiatimes.com//entertainment/events/lucknow/yashpal-sharma-and-mita-vashisht-shoot-for-a-holi-song-in-lucknow/articleshow/62724857.cms',
 'http://timesofindia.indiatimes.com//entertainment/tamil/movies/news/kumararaja-sir-is-a-perfectionist-mrinalini-ravi/articleshow/62725081.cms',
 'http://timesofindia.indiatimes.com//city/kochi/ranjith-sankar-asks-how-long-will-the-common-man-have-to-bear-the-brunt-of-increasing-fuel-prices/articleshow/62724906.cms',
 'http://timesofindia.indiatimes.com//entertainment/kannada/movies/news/yashwanth-shetty-wants-to-be-a-villain-with-a-difference/articleshow/62724330.cms',
 'http://timesofindia.indiatimes.com//city/kochi/kerala-blasters-players-including-more-local-talent-will-bring-in-more-fans/articleshow/62724678.cms',
 'http://timesofindia.indiatimes.

In [20]:
#creating empty list and storing data from each news article by this code

Date=[]
Author=[]
Vertical=[]
Headline=[]
Description=[]
fina=final

num=0
for f in fina:
    num=num+1
    print(num,"executing ...")
    driver.get(f)
    time.sleep(10)
    
    try:
        try:
            h=driver.find_element_by_xpath("//div[@class='as_heading']/h1/arttitle")
            Headline.append(h.text.replace("\n",''))
        
        except:
            h=driver.find_element_by_xpath("//h1[@class='_1Y-96'][1]")
            Headline.append(h.text.replace("\n",''))
            
    
    except NoSuchElementException :
        Headline.append("-")
        
        
    try:
        try:
            d=driver.find_element_by_xpath("//div[@data-plugin='dateformat']")
            Date.append(d.text.replace("Created: ",''))
        
        except:
            d=driver.find_element_by_xpath("//div[@class='yYIu- byline'][1]/span")
            Date.append(d.text.replace("Updated: ",''))
            
    
    except NoSuchElementException :
        Date.append("-")
        
        
    try:
        try:
            a=driver.find_element_by_xpath("//div[@itemprop='author']/a")
            Author.append(a.text)
        
        except:
            a=driver.find_element_by_xpath("//div[@class='yYIu- byline'][1]/a")
            Author.append(a.text)
            
    
    except NoSuchElementException :
        Author.append("-")
        
    try:
        try:
            v=driver.find_element_by_xpath("//div[@class='wrapper']/ol/li[2]/a/span")
            Vertical.append(v.text.replace("\n",''))
        
        except:
            v=driver.find_element_by_xpath("//div[@class='_1nPcO']/ul/li[2]/a/span")
            Vertical.append(v.text.replace("\n",''))
            
    
    except NoSuchElementException :
        Vertical.append("-")
    
    try:
        try:
            des=driver.find_element_by_xpath("//div[@class='Normal'][1]")
            Description.append(des.text.replace("\n",''))
        
        except:
            des=driver.find_element_by_xpath("//div[@class='_3YYSt clearfix  '][1]")
            Description.append(des.text.replace("\n",''))
            
    
    except NoSuchElementException :
        Description.append("-")
        
    

1 executing ...
2 executing ...
3 executing ...
4 executing ...
5 executing ...
6 executing ...
7 executing ...
8 executing ...
9 executing ...
10 executing ...
11 executing ...
12 executing ...
13 executing ...
14 executing ...
15 executing ...
16 executing ...
17 executing ...
18 executing ...
19 executing ...
20 executing ...
21 executing ...
22 executing ...
23 executing ...
24 executing ...
25 executing ...
26 executing ...
27 executing ...
28 executing ...
29 executing ...
30 executing ...
31 executing ...
32 executing ...
33 executing ...
34 executing ...
35 executing ...
36 executing ...
37 executing ...
38 executing ...
39 executing ...
40 executing ...
41 executing ...
42 executing ...
43 executing ...
44 executing ...
45 executing ...
46 executing ...
47 executing ...
48 executing ...
49 executing ...
50 executing ...
51 executing ...
52 executing ...
53 executing ...
54 executing ...
55 executing ...
56 executing ...
57 executing ...
58 executing ...
59 executing ...
60 exe

WebDriverException: Message: unknown error: net::ERR_CONNECTION_RESET
  (Session info: chrome=92.0.4515.159)


In [203]:
Author

['Deepa Soman',
 '-',
 'Sharanya CR',
 '-',
 'Prathibha Joy',
 'Sanjith Sidhardhan',
 'Dhwani Desai',
 '-',
 '-',
 'M Suganth',
 'Sindhu Vijayakumar',
 'Deepika Jayaram',
 'Anjana George',
 '-',
 '-',
 'Prathibha Joy',
 'V Lakshmi',
 'Dhwani Desai',
 'Dhwani Desai',
 '-',
 '-',
 '-',
 '-',
 'Aswin J Kumar',
 '-',
 'M T Saju',
 'Anjaya Anparthi',
 'Marcus Mergulhao',
 'Reema Gowalla',
 'Reema Gowalla',
 'Umesh Isalkar',
 'Jisha Surya',
 'Anjaya Anparthi',
 '-',
 '-',
 '-',
 'Anjaya Anparthi',
 '-',
 'A Selvaraj',
 '-',
 '-',
 '-',
 'Arsh Behal',
 '-',
 '-',
 '-',
 '-',
 '-',
 'Karishma Kotwal',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 'Rajat Pandit',
 '-',
 'Shiladitya Pandit',
 'Joychen Joseph',
 '-',
 '-',
 'Ashish Mehta',
 '-',
 '-',
 '-',
 '-',
 '-',
 'Rosamma Thomas',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 'Stanley Pinto',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 '-',
 'Niharika Lal',
 'Niharika Lal',
 

In [204]:
Vertical

['CITY NEWS',
 'Entertainment',
 'Entertainment',
 'CITY NEWS',
 'Entertainment',
 'CITY NEWS',
 'Entertainment',
 'CITY NEWS',
 'Entertainment',
 'Entertainment',
 'Entertainment',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'Entertainment',
 'Entertainment',
 'Entertainment',
 'Entertainment',
 '-',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'Entertainment',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'Entertainment',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 '-',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'Entertainment',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'CITY NEWS',
 'INDIA NEWS',
 'Entertainment',
 'CITY NEWS',
 'CITY NEWS',
 'C

In [205]:
Headline

['Celebrity Reporter: Actress Ahaana Krishna writes about the need for including moral science in schools',
 'Yashpal Sharma and Mita Vashisht shoot for a Holi song in Lucknow',
 'Kumararaja sir is a perfectionist: Mrinalini Ravi',
 'Ranjith Sankar asks: How long will the common man have to bear the brunt of increasing fuel prices?',
 'Yashwanth Shetty wants to be a villain with a difference',
 'Kerala Blasters players: Including more local talent will bring in more fans',
 'There wasn’t even the R of romance between me and JK',
 'Meet Shruti Huddar — the BSF daredevil from Belagavi',
 'I went to Bigg Boss with the sole aim of surviving for 98 days: Chandan Shetty',
 'VJ Aadhavan gives voice-over for a golden retriever',
 'Gender doesn’t matter in this profession: Kiruthiga Udhayanidhi',
 'Govind Menon on taking the less trodden path to independant music',
 'Celebrity Reporter: Actor Jayasurya finds out why is humanity missing in our society?',
 'More trains, better stations on passeng

In [199]:
len(Description)==len(Author)==len(Headline)==len(Vertical)==len(Date)

True

In [206]:
Description

['Intolerance of various kinds is on the rise all around us, online and offline. From mere bullying to taking someone’s life for the silliest of reasons, it takes many shapes and forms and Kerala isn’t an exception. Often, the easily quoted solution for this is that one should ‘teach kids to behave’ right from a young age, at homes. I’m not saying we can forget the family’s role in the process, but it’s at schools that they spent the maximum amount of time. Are there effective mechanisms in our curriculum today which teach kids to be humane citizens?Every school time table has an hour dedicated for moral science in which children are supposed to be imparted ethics, values and life lessons. But the period isn’t given the same importance as physics, math or chemistry, and often ends up being a namesake. I spoke to Anita Krishnan, a senior faculty member of Holy Angels ISC School, Thiruvananthapuram, who agrees that we need to take up the subject seriously. “Education’s primary aim should

In [30]:
#creating a dataframe for scrapped data

df=pd.DataFrame({
"Date":Date,    
"Author":Author,
"Vertical":Vertical,
"Headline":Headline,
"Description":Description})

df.to_csv("Sample TI part 1.csv")

In [31]:
df

Date         Author       Vertical  \
0     Feb 1, 2018, 00:00 IST    Deepa Soman      CITY NEWS   
1     Feb 1, 2018, 00:00 IST              -  Entertainment   
2     Feb 1, 2018, 00:00 IST    Sharanya CR  Entertainment   
3     Feb 1, 2018, 00:00 IST              -      CITY NEWS   
4     Feb 1, 2018, 00:00 IST  Prathibha Joy  Entertainment   
...                      ...            ...            ...   
1256  Feb 3, 2018, 04:41 IST              -      CITY NEWS   
1257  Feb 3, 2018, 04:41 IST              -      CITY NEWS   
1258  Feb 3, 2018, 04:42 IST   NAOMI CANTON     WORLD NEWS   
1259  Feb 3, 2018, 04:42 IST              -      CITY NEWS   
1260  Feb 3, 2018, 04:42 IST      Ahmed Ali      CITY NEWS   

                                               Headline  \
0     Celebrity Reporter: Actress Ahaana Krishna wri...   
1     Yashpal Sharma and Mita Vashisht shoot for a H...   
2     Kumararaja sir is a perfectionist: Mrinalini Ravi   
3     Ranjith Sankar asks: How long will the common ...   
4     Yashwanth Shetty wants to be a villain with a ...   
...                                                 ...   
1256          Two hawkers killed under train at Palghar   
1257     Constable arrested for abetting wife’s suicide   
1258             UK mosque attacker jailed for 43 years   
1259  IIT-Bombay old boy gives Rs 1 crore for studen...   
1260   Couple given 5 years for duping US job aspirants   

                                            Description  
0     Intolerance of various kinds is on the rise al...  
1     By 'Abhishek.Shukla2@timesgroup.comActors Yash...  
2     Recently, Mrinalini Ravi, who shot to fame wit...  
3     Why are diesel prices almost as high as petrol...  
4     A National School of Drama alumnus who taught ...  
...                                                 ...  
1256  MUMBAI: Two unidentified hawkers died under a ...  
1257  MUMBAI: A constable from the Gholvad police st...  
1258  LONDON: A British right-wing terrorist who ram...  
1259  MUMBAI: The IIT tuition fee increase in 2016 h...  
1260  MUMBAI: A couple has been sentenced to five ye...  

[1261 rows x 5 columns]

#### This is the final data available.

#### Note: This is only first part(feb1-feb 5) of code where only 1261 rows are scrapped. I made multiple copies of this code and splitted the news link using slice for scrapping the data faster. Since i don't have efficient power and internet suppy i have to follow this. I have got and combined all csv to get the final output file which is shared with you.